### ISSUE TITLE
[Cohort-1 TASK 3] Eliminate redundant zero-KL reference computation

### ISSUE THREAD

**Issue by g815213795** [redai-infra/Relax#99]
## 1. Scope

I claimed Task 03 (Text Dense Performance Optimization) in #86.

The experiment will use the Qwen3-4B fully asynchronous GRPO recipe. The goal is to improve end-to-end throughput and GPU utilization without changing the model, dataset, effective batch size, sampling configuration, sequence limits, precision, reward, or training objective.

## 2. Repository and Baseline

- Repository: https://github.com/redai-infra/Relax
- Upstream commit: `30f219a1c963137129570dd5485a9c3d82390548`
- Base recipe: `scripts/training/text/run-qwen3-4B-4xgpu-async.sh`
- Model: `Qwen/Qwen3-4B`
- Training data: `zhuzilin/dapo-math-17k`
- Evaluation data: `zhuzilin/aime-2024`

## 3. Environment

- GPU: 4× NVIDIA RTX 5090 32 GB
- CPU: 100 vCPU
- RAM: 360 GB
- OS: Ubuntu 22.04
- Python: 3.12.3
- PyTorch: 2.11.0+cu130
- CUDA: 13.0
- Ray: 2.56.1
- SGLang: 0.5.12.post1

Baseline and optimized runs will use the same environment and hardware.

## 4. Fixed Workload

The effective workload will remain unchanged:

```bash
--rollout-batch-size 4
--n-samples-per-prompt 8
--rollout-max-response-len 2048
--global-batch-size 8
--num-rollout 200
--max-staleness 1
--num-iters-per-train-update 1
--fully-async
```

Launch command:

```bash
MODEL_DIR=/data/assets \
DATA_DIR=/data/assets \
EXP_DIR=/data/runs/<baseline-or-optimized> \
NUM_ROLLOUT=200 \
bash scripts/training/text/run-qwen3-4B-4xgpu-async.sh
```

## 5. Proposed Optimizations

Preliminary profiling shows that the actor policy-loss path computes full-vocabulary entropy even though:

```bash
--entropy-coef 0.00
```

It also creates large FP32 logits clones during log-probability computation, causing high peak memory and preliminary OOMs.

Planned changes:

- Skip entropy computation when its coefficient is zero and no diagnostic requires it.
- Compute FP32 policy log-probabilities in bounded token chunks.
- Avoid unnecessary full-vocabulary logits clones.
- Preserve existing behavior for non-zero entropy and KL configurations.
- Add focused output and gradient-equivalence tests.

## 6. Evaluation and Deliverables

Baseline and optimized versions will each run for 200 steps. Each run will be divided into at least three identical post-warm-up statistical windows.

I will report:

- End-to-end response tokens/s
- Step-time mean and P50/P95
- Rollout, log-prob, actor-train, and weight-update time
- Average GPU utilization and peak VRAM
- Loss, reward, KL, and response-length statistics
- OOM, NaN/Inf, and dropped-sample status
- Full commands, raw logs, environment records, and test results

Target: at least 5% higher primary throughput, with no OOM, dropped samples, numerical instability, or quality regression.
